### ISSUE TITLE
Footnote precedes a reference to it

### ISSUE THREAD

**Issue by jsx97** [typst/typst#5314]
### Description

Sometimes a footnote precedes a reference to it. For example:

```typ
= Lorem
#lorem(580)

== Ipsum
#lorem(10)#footnote[#lorem(20)] #lorem(2000)
```

Here, the reference to a footnote is located at Page 2, but the footnote itself is located at Page 1. As far as I know, a footnote should never precedes a reference to it. And so, if I understand correctly, this is a bug.

<img width="811" alt="image" src="https://github.com/user-attachments/assets/ad36cbc3-bb05-4aaa-9d6d-c3b455a330e7">

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
Yes, this is a bug. In complex scenarios, it is a bit hard (but possible, just not yet implemented) to avoid, but this case is rather simple, so it should definitely work.

**Comment by pauladam94** [comment]
Similarly this code :
```typst
#v(96%)

= Header #footnote[footnote]
```
Gives up this image :
![image](https://github.com/user-attachments/assets/4d6b559b-730a-43d6-802f-aef5dda67af2)

It is reverse to the original example but similar issue : the footnote is present is at the next page of the position of the reference.

**Comment by PgBiel** [comment]
> Similarly this code :
> (...)
> It is reverse to the original example but similar issue : the footnote is present is at the next page of the position of the reference.

This is different as that's related to footnotes' migration behavior, where the element where the footnote originally appeared is moved to the next page / column to ensure it appears together with the footnote entry, but there are times where (due to lack of space at the bottom, e.g. footnote entry is too long) this isn't possible. Still, your case seems simple enough that it should work - but I'd open a new issue for separate investigation.

**Comment by PgBiel** [comment]
I think this is a regression as the example works fine on 0.11, so I will mark it as such.

**Comment by PgBiel** [comment]
Minimal reproduction:

```typ
#set page(width: 10cm, height: 3cm, margin: 0.8cm)

Foot #footnote[foot] \ foot
```

![image](https://github.com/user-attachments/assets/a33a53bb-cfed-4b41-9360-9e45648d1335)

I think it's related to orphan prevention; the first line fits in the first page, but moves to the second page due to `\ foot` (removing it keeps both the first line and the footnote in the same page).

**Comment by PgBiel** [comment]
Alright, I've managed to debug it a bit. Here's what's happening:

1. **Without** the footnote, both lines (/ the first two) fit, not triggering orphan prevention:
     ```typ
     #set page(width: 10cm, height: 3cm, margin: 0.8cm)

     Foot \ foot
     ```
     
     ![Just one page: "Foot" in one line and "foot" in the next](https://github.com/user-attachments/assets/51ff3b5f-ea9b-4b92-ba92-e2c27533d271)

2. Therefore, on the very first layout pass, as there are no known footnotes yet, both lines are laid out normally as separate frames, as they both fit. This calls `distributor.frame` which searches for footnotes in the lines, adding them to be placed as part of column insertions. This triggers a relayout.
3. The relayout triggered by the footnote **shrinks the available space** in the page by the **height of the footnote entries section**. This triggers the second line to move to the next page, which in turn triggers the first to move as well by orphan prevention - this triggers another relayout.
4. However, in this second relayout of the first page, despite both lines being moved to the next, **the footnote is still pending to be placed** in the first page, as it was never removed from column insertions, and so it is placed.
5. The two lines are then placed in the next page normally.

We can see that, ignoring the timeline of events, the main problem is that placing the lines is what's causing them to be moved in the first place; moving them to the next page wouldn't trigger this need, but then they wouldn't have had to be moved at all...

The minimal case I'm showing, in fact, is actually pathological, in the sense that text + footnotes _cannot_ fit in one page, so the footnote variant cannot be upheld either way. But 0.11.1 arguably handles it better: the footnote is placed on the same page (which is weird; it should be moved), though it "fits" by not fitting (the first frame is empty and the text is automatically moved due to lack of space).

![footnote appears empty in the first page and then footnote text shows up in the next page](https://github.com/user-attachments/assets/eaa70bc9-d69f-4a09-bd4b-fc7f60568b76)

Even then, 0.12 exhibits the same problem for non-pathological cases, as shown.

```typ
#set page(width: 10cm, height: 5.3cm, margin: 0.8cm)

Foots

Foots

Foots

Foot #footnote[foot] \ foot
```

0.12:
![footnote appears on the first page with just "Foots" x3; Foot \ foot appears on the second page](https://github.com/user-attachments/assets/3fe89d34-42d0-41c2-b3ff-6a3f09c1b45b)

0.11:
![footnote and Foot \ foot are both moved to the second page. First page only contains "Foots" x3](https://github.com/user-attachments/assets/f5f05647-754b-42fd-a2c4-e4ee53833df1)

**Comment by PgBiel** [comment]
With that conclusion, we can find some other interesting cases:

```typ
#set page(width: 10cm, height: 6.3cm, margin: 0.8cm)

#box(height: 0cm)[a]

= a #footnote[abc]

= b #footnote[dominoes]

Foot #footnote[foot] \ foot
```

0.12:

![all 3 footnotes appear on the first page, but the headings and paragraph appear on the second](https://github.com/user-attachments/assets/2359d35d-ae29-402e-a773-5d7f10d2c8d6)

0.11 (no stickiness so it's not as interesting...)

![the headings and their footnotes appear on the first page, the paragraph and its footnote appear on the second](https://github.com/user-attachments/assets/f98729b6-8cbf-418a-b4f4-30213ff7de8c)

**Comment by PgBiel** [comment]
In fact, we no longer have to blame orphan prevention for this... sticky blocks are also affected in general. Main problem seems to be that we "remember" footnotes of moved things.

```typ
#set page(width: 10cm, height: 6.3cm, margin: 0.8cm)

#box(height: 0cm)[a]

= a #footnote[abc]

= b #footnote[dominoes]

#block(breakable: false, height: 3cm)
```

![headings are in the second page, footnotes stayed on the first](https://github.com/user-attachments/assets/0e677029-a03a-4d56-953a-272024926afb)

**Comment by laurmaedje** [comment]
That all makes sense. I was aware of the possibility of this happening, just didn't realize that widow/orphan prevention would trigger it. :/

**Comment by hurzl** [comment]
Does this have to with why I get a widow with:

```
#lorem(586)

#lorem(60)#footnote[södlkfjsldjf]
```

**Comment by demisjohn** [comment]
Is there any **workaround** for this, while you all continue to work on fixing the bug?

Anything we can do in-text to fix this (with no compiler/interpreter changes)?  Any suggestions would be greatly appreciated, as I can find no current solution apart from adding a sentence of fluff to my content!

Thank you kindly.

PS. in case it's useful, here is a Typst forum discussion about this same issue: https://forum.typst.app/t/how-can-one-avoid-a-footnote-appearing-on-the-page-before-its-anchor-in-the-text/5182/2

**Comment by demisjohn** [comment]
I'd like to point out that (a) this was reported 2 years ago in 2024 and apparently is still unresolved, and (b) there is apparently _no_ workaround as users await a bugfix.  Either a bugfix _or_ a workaround would be acceptable from a usage standpoint.

### Current workaround:
I will add 2 lines of fluff/extra text prior to the in-text superscript to fix the reflow. This of course requires me to check every footnote and manually reflow the text if this error is occurring (although I think it is only occurring on one of the ~7 footnotes in the book).

**Is there any way to escalate this issue?**  I do apologize that my own programming prowess is inadequate to this task, or else I would attempt to contribute myself.

**Comment by trespaul** [comment]
I really don't like to or want to make demands of open source projects, but I agree that this should be somewhere high up on the priorities list as a prerequisite for being taken seriously as a professional typesetting system. I typeset a volume of an academic journal in Typst last year, and the footnotes were the only thing that I was dissatisfied with in the final product.

@demisjohn something I tried was to add extra spacing between paragraphs or before headings (which looks a bit more natural/excusable) in the surrounding text — have you tried that as a workaround?

**Comment by cady-b** [comment]
Hi everyone! Please know that there is awareness on this being an issue, and it's not stale for a lack of care. After the recent comments here, I [brought it up](https://discord.com/channels/1054443721975922748/1054443722592497796/1510746268736163911) among other contributors on [the Discord](https://typst.app/open-source#support) again, asking if tackling this soon has gotten any more feasible.

Unfortunately, (to my understanding of the situation; other contributors / the team are more versed here but they don't have as much time 😅) the sentiment is still that:
- As PgBiel analyzed above, a proper fix would involve comprehensive changes to the layout system. As far as the people I talked to could tell me, nothing happened since that comment which would make the patch trivial.
- The ambition is to sustainably work out footnote placement (and that of other elements with more complicated positioning), but what we're seeing here plays into many fundamental parts of how we approach layout. Typst is conceptually well positioned to excel at this, but building up to there from abstractions over our primitives is a gradual process that is still in motion. There is a sort of order to which sets of problems are solved; unfortunately we can't really skip around in it for individual high level issues like this one. See also: https://laurmaedje.github.io/posts/layout-models
- Thus I also believe quite a lot of consideration will have to be put towards this being compatible with other planned changes (or indeed past ones, given this being a regression), and to not put undue burden on maintainers in the meantime. This makes the issue infeasible for people new to the project to resolve, and even with experienced contributors, the team would have to be involved quite a bit.
- Since the issue is rare overall and doesn't fully block users from getting their documents delivered, it's not feasibly to escalate this to `urgent` again currently (which would block releases from happening). Do note that beyond day-to-day upkeep, the team and contributors are already very hard at work fixing issues (see i.e. PgBiel's [work on list layout](https://redirect.github.com/typst/typst/pull/7895), or Laurenz one day [resolving long-standing crashes in SVG](https://redirect.github.com/typst/typst/pull/7766) which I was hunting down for months) and further setting Typst apart as a truly competent package (see i.e. the ongoing accessibility work, close attention to the PDF spec via [Krilla](https://github.com/LaurenzV/krilla), [Interactive Forms](https://bsky.app/profile/typst.app/post/3mmrma5rjss2u), towards HTML export with [MathML](https://redirect.github.com/typst/typst/pull/7436) and the basis for more intricate [styling](https://redirect.github.com/typst/typst/pull/8178), allowing for [multiple bibliographies](https://redirect.github.com/typst/typst/pull/8307) and renewed attention to CSL compliance in [hayagriva](https://github.com/typst/hayagriva), etc.)

Also: Typst scripting plain does not expose the tools needed to provide a workaround, short of embedding another layout engine.

But: Further improvements to layout [are on the roadmap](https://typst.app/docs/roadmap/#:~:text=Layout,-Fix), and this is being considered while we build up to that. It's a hard ask dedicating a lot of time to this specific fix at the current time, but we're on track for it to fall into place more nicely in the future.

---

Though I am very happy to report that our dear @ensko has envisioned a way to at least _detect_ the error has occurred.

We're missing a nicer warning system still, but together we were able to conceive of a way to notify the user such that they can take appropriate action as described in above comments, exactly when they need to output good layout (without worrying about it throughout the rest of the process):

```typ
// warn on https://github.com/typst/typst/issues/5314
#show footnote.entry: it => {
  let _warn(body) = {
    let footnote-placement-error = [#label(body)]
  }

  let (my-location, my-page) = (it.location(), it.location().page())
  let (note-location, note-page) = (it.note.location(), it.note.location().page())

  // identify possible earlier references via `#footnote(<a>) #footnote[Foo]<a>`
  let label-notes = query(footnote).filter(v => type(v.body) == label)
  let earlier-reference = label-notes.any(v => {
    let candidate = locate(v.body)
    candidate == note-location and v.location().page() <= my-page
  })

  if note-page > my-page and not earlier-reference {
    let n = numbering(it.note.numbering, ..counter(footnote).at(my-location))
    _warn("Footnote " + n + " was produced before its first reference")
  }

  it
}
```

It will surface like this with the webapp:

<img width="400" src="https://github.com/user-attachments/assets/3e1c1545-9354-4b15-9ece-65b0f9b7a849" />

and like this with the cli:

```
warning: label `label("Footnote 1 was produced before it's first reference")` is not attached to anything
  ┌─ a.typ:3:37
  │
3 │     let footnote-placement-error = [#label(body)]
  │                                      ^^^^^^^^^^^
```

As far as I can tell, the approach takes one extra layout iteration (see [this blog post](https://blog.ensko.at/2025/07/07/layout-convergence/) for a nice intro no that) to resolve in the basic case, and I've seen it go up to two in more complex scenarios. As such, it should work in most documents. Though if you're already close to the limit and your footnotes are blocked on introspection, convergence warnings could still come up; feel free to ask on the [Forum or Discord](https://typst.app/open-source#support) for support should that happen after applying the rule!

**Comment by trespaul** [comment]
Thank you for the update @cady-b! I just want to reiterate how much I (and the rest of us, I presume) appreciate the contributors' efforts. I'm very excited for the future of Typst!

**Comment by ensko** [comment]
I had an additional thought that may be useful in connection to the detection workaround: you can use
```typ
#place(bottom, float: true, clearance: 0pt, v(1cm))
```
to reduce the space available for text content on a page by some value (here 1cm). That would (hopefully) affect the layout so that the footnote naturally appears on the next page, ideally not actually shifting any content (because it was already on the next page due to stickiness, orphan prevention, etc.)

It's another manual step, but may help in cases where adding fluff text seems worse than empty space.

**Comment by demisjohn** [comment]
Thanks @cady-b , thanks for the warning/detection code, that is very helpful!  Also it's good to know (a) there isn't anything we can do to coerce the layout engine to fix this (saves me time trying to mess with that!), and (b) the team is aware.  Much appreciated, both the update and reminding the team that the issue is still open.

@trespaul thanks for that - you're right that adding `#v(1pt)` or similar on headings on the same page works very well as a workaround, I don't think a reader would ever notice. For someone that doesn't have a heading on the same page, you could probably add the vertical space between paragraphs. Thanks for the suggestion!

(I will add that I started using Typst about 2 months ago, and will never go back to LaTeX!  It is really a great system.)

**Comment by pauladam94** [comment]
@cady-b the warning inside Typst code is indeed very useful. Does the Typst compiler check for this kind of behavior inside his test suite ? It should one of the test that is not a visual vérification or a verification that the output is exactly the same that could be very handy.

**Comment by cady-b** [comment]
Well, as this is currently broken, there is no test in place at the moment 😅

Though it's standard practice to add tests in PRs that fix issues (to avoid them regressing again). I'm sure that will be the case here as well! :)

**Comment by laurmaedje** [comment]
Another reproducer from https://github.com/typst/typst/issues/8603:

```typ
#set page(paper: "a6")

Lucas goes to school every day of the week. He has many subjects to go to each school day: English, art, science, mathematics, gym, and history. His mother packs a big backpack full of books and lunch for Lucas.

His first class is English, and he likes that teacher very much. His English teacher says that he is a good pupil, which Lucas knows means that she thinks he is a good student.

His next class is art. He draws on paper with crayons and pencils and sometimes uses a ruler. Lucas likes art. It is his favorite class.

His third class is science. This class is very hard for Lucas to figure out, but he gets to work with his classmates a lot, which he likes to do. His friend, Kyle, works with Lucas in science class, and they have fun.

Then Lucas gets his break for lunch. He sits with Kyle while he eats. The principal, or the headmaster as some call him, likes to walk around and talk to students#footnote[it's just the seargent] during lunch to check that they are behaving.

The next class is mathematics, which most of the students just call mathall#footnote[let it be]. Kyle has trouble getting a good grade in mathematics, but the teacher is very nice and helpful.

His fourth class is gym. It is just exercising.

History is his last class of the day. Lucas has a hard time staying awake. Many lessons are boring, and he is very tired after doing gym.
```

**Comment by laurmaedje** [comment]
Yet another case; from https://github.com/typst/typst/issues/8636:
```typ
#lorem(5)
#footnote(lorem(1100))
#lorem(300)
```

**Comment by demisjohn** [comment]
@cady-b any chance this very useful _Warning_ could be implemented in the compiler itself?  For example, so VSCode can report the possible error.  I don't know if the free compiler does actually have the ability to pop warnings in VS Code... but that is my primary use-case.  Thanks!

**Comment by cady-b** [comment]
Hi demis! The compiler can emit warnings, but to see them in the editor, one also needs an LSP server that bridges the two; one such for Typst would be [tinymist](https://github.com/Myriad-Dreamin/tinymist/). With that, the snippet above should also work in VSCode as is:
<img width="1340" height="155" alt="Image" src="https://github.com/user-attachments/assets/d1a29ede-79e9-47f9-a2fc-b5202490bcec" />

In terms of integrating a warning directly without having to paste the snippet, I guess it'd mostly depend on how disruptive such a change would be to the rest of the compiler. On its own, I'd definitely consider it a net benefit (though the one you'd have to convince is laurmaedje); if someone has the time to investigate, that'd be nice!

If it turns out to not be a great fit, a [lint](https://myriad-dreamin.github.io/tinymist/feature/linting.html) could also be proposed to the tinymist project. It has a number of warnings separate from the compiler.

**Comment by demisjohn** [comment]
Yes I am using Tinymist - it is a very good experience.  Adding the warning code to Tinymist is a good idea, if that would be less disruptive/risky. 

Since the bug is in Typst, not Tinymist, do you think the Tinymist project would be open to adding that warning? (Should be ok I suppose if some of their existing warnings operate like this.)

**Comment by cady-b** [comment]
Adding the warning in Typst itself would benefit more people and should thus be the first avenue. Tinymist might be an alternative in case a solution is found but doesn't get accepted into the compiler for whatever reason.

I don't know how their processes for lints look like unfortunately, just wanted to note it as another option.
